text/go/0001-go-bom.md

```diff
@@ -0,0 +1,164 @@
+
+1. `app`: Include only those modules that the target application actually depends on. Modules required by tests or
+   packages that are not imported by the application are not included. Build constraints are evaluated, which enables a
+   very detailed view of what's really compiled into an application's binary.
+2. `mod`: Include the aggregate of modules required by all packages in the target module. This optionally includes
+   modules required by tests and test packages. Build constraints are NOT evaluated, allowing for a "whole picture" view
+   on the target module's dependencies.
+3. `bin`: Offers support of generating rudimentary SBOMs from binaries built with Go modules.
+
+Of these, we are only interested in `app`. The buildpacks do not currently support building modules, or libraries, which
+is the primary use-case for `mod`. The `bin` command is for pre-built executables, which is unnecessary for the
+buildpack as it has access to the source code. The BOM created by the `bin` command also contains less information
+than the one created via the `app` command.
+
+The output BOM contains the following fields for each module:
+
+* name,
+* version,
+* hash,
+* source URI,
+* license ID,
+* and package URL
+
+Note: The `license` field can be retrieved when the flag `-licenses` is passed to the command.
+
+Almost all of fields that we aim to support
+per [the Paketo BOM RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0033-bill-of-materials.md#overall-schema)
+are available. The only one that is not available is `description`, and this is because there is no standard way to
+capture this information in Go projects (as compared with other languages like nodejs and rust, where the package
+description files contain a field for this metadata). See unresolved questions
+below.
+
+In order to support this tool in both the online and offline cases, we will host
+the binary tool on the [dep-server](https://github.com/paketo-buildpacks/dep-server).
+
+### Performance
+
+The addition of this tool to the language family order groupings might have an impactful change to the overall
+performance of the buildpacks.
+
+Performance metrics are still TBD, but for a typical project (e.g. `pack`) we observe that it takes about 10-30 seconds
+to construct a BOM.
+
+We can investigate ways to improve this performance.
+
+### Language Family Additions
+
+Using the CycloneDX tool to generate the BOM will involve some changes to the Go language family buildpack order
+groups as follows:
+
+* A single BOM generation buildpack will be added. It's primary functions are:
+    * perform the BOM dependency tool retrieval from the dep-server,
+    * execute BOM generation with the tool (`cyclonedx-gomod app -o bom.json`), and
+    * contribute the BOM to the `launch.toml` `[bom]` label.
+
+* For the Go Mod buildpack order group, the BOM generation buildpack will run directly before the `go-build` buildpack.
```

Comment:
Is there any particular reason this buildpack runs before go-build and not after? In my head, I think of doing accounting and writing a Bill of Materials as a last step one would do after all the build commands are completed.

Just for clarity, it would be nice if you can print out the entire order group here

Reply:
No particular reason, except the BOM process doesn't require the build process to have run. We modeled this ordering off the nodejs BOM buildpack, and this is the analogous ordering to what that language family does.

We're happy to change the ordering if folks prefer that.

Reply:
This order grouping makes sense to me logically since the BOM creation directly will need the Go modules. +1 to getting the full order group in the RFC.

Reply:
@sophiewigmore which ordering do you prefer? The one that is written, or the proposed change? Either is fine by us.

Happy to add the full ordering here for clarity.

Reply:
You would need to generate the BOM before the `go-build` buildpack runs because it removes all of the source from the working directory: https://github.com/paketo-buildpacks/go-build/blob/c40e20b61d57bed68095055814b10b74f4b1e617/build.go#L87-L90

After it executes, there won't be any `go.mod` or `go.sum` files to help with generating the BOM.